taichi/analysis/gather_uniquely_accessed_pointers.cpp

```diff
@@ -55,6 +59,18 @@ class LoopUniqueStmtSearcher : public BasicStmtVisitor {
     }
   }
 
+  void visit(DecorationStmt *stmt) override {
+    if (stmt->decoration.size() == 2 &&
+        stmt->decoration[0] ==
+            uint32_t(DecorationStmt::Decoration::loop_unique)) {
+      if (loop_unique_.find(stmt->operand) == loop_unique_.end()) {
+        loop_unique_[stmt->operand] = stmt->decoration[1];
+        num_different_loop_indices = std::max(loop_unique_[stmt->operand] + 1,
+                                              num_different_loop_indices);
```

Comment:
Could you please give me an example that `loop_unique_[stmt->operand]` is greater than or equal to `num_different_loop_indices` before this statement? I feel a bit worried if such a thing can happen because this analysis should guarantee to produce a correct result even if *some* `DecorationStmt`s are removed.

Reply:
This is with the case where the lowering pass has transformed a 2d or 3d range / struct on `ti.any_arr` to a linear index, and generated IR that calculates the ND index. In those case the offloaded stmt use a 1d index, but the original code written by the user is ND.

Reply:
I see. So I wonder if we still have `LoopIndexStmt` in this offloaded stmt using the 1d index? If so, this line https://github.com/taichi-dev/taichi/blob/c231aa3b482672a7f37e781df286380ce6384fd5/taichi/analysis/gather_uniquely_accessed_pointers.cpp#L34 may not be able to detect the uniqueness properly -- note that `num_different_loop_indices` is a variable in this class used both in `is_ptr_indices_loop_unique(GlobalPtrStmt *stmt)` and `is_ptr_indices_loop_unique(ExternalPtrStmt *stmt)`.

And, what if we have multiple ND-arrays with different numbers of dimensions in one offloaded stmt? Maybe we can refactor the `num_different_loop_indices` used here (we may need to rename this variable and keep the original one equal to the original dimensions of the offload) into a map from the external arrays to the number of dimensions.

We may need to add some notes that the `DecorationStmt` is essential for alias analysis and cannot be removed until no more alias analysis is going to be performed (e.g., in codegen phase).

Reply:
I'm relying on this decoration only being added for loops that we transformed. If different dimensions of NDarrays being used in the loop, the loop itself is essentially only a certain amount of dimensions. The frontend statement does not ever use the linearized index in any ways, and the backend IR neither will, thus I don't think it will cause a problem. If these statements are removed, alias analysis will naturally not happen because the linearized index statement is never used in any way.

Reply:
I see, so IIUC there is only one set of decoration statements in one loop, and only if we are looping over an NDarray, right?

We should add some comments at
https://github.com/taichi-dev/taichi/blob/c231aa3b482672a7f37e781df286380ce6384fd5/taichi/analysis/gather_uniquely_accessed_pointers.cpp#L189
 that we are initializing `num_different_loop_indices` here in the NDarray case. It can be confusing because we initialize this variable at two places that are far away.